geomstats/learning/riemannian_mean_shift.py

```diff
@@ -0,0 +1,147 @@
+        self.metric = metric
+        self.bandwidth = bandwidth
+        self.tol = tol
+        self.mean = FrechetMean(self.metric, **FrechetMean_kwargs)
+        self.centers = None
+
+    def __intersets_distances(self, points_A, points_B, n_jobs=1, **joblib_kwargs):
+        """
+        Parallel computation of distances between two sets of points.
+        """
+        n_A, n_B = points_A.shape[0], points_B.shape[0]
+
+        @joblib.delayed
+        @joblib.wrap_non_picklable_objects
+        def pickable_dist(x, y):
+            return self.metric.dist(x, y)
+
+        pool = joblib.Parallel(n_jobs=n_jobs, **joblib_kwargs)
+        out = pool(
+            pickable_dist(points_A[i, :], points_B[j, :])
+            for i in range(n_A)
+            for j in range(n_B)
+        )
+
+        # assuming numpy backend
+        return np.array(out).reshape((n_A, n_B))
+
+    def fit(
+        self,
+        points,
+        n_centers,
+        n_jobs=1,
+        max_iter=100,
+        init_centers="from_points",
+        kernel="flat",
+        **joblib_kwargs
+    ):
+        """
+        Fit centers in all the input points to find 'n_centers' number of clusters.
+        """
+
+        @joblib.delayed
+        @joblib.wrap_non_picklable_objects
+        def pickable_mean(points, weights):
+            return self.mean.fit(points, weights=weights).estimate_
+
+        # 'from_points' is the default initialization for centers used
+        if init_centers == "from_points":
+            n_points = points.shape[0]
+            centers = points[np.random.randint(n_points, size=n_centers), :]
+        if init_centers == "random_uniform":
+            n_centers = centers.shape[0]
+            centers = self.manifold.random_uniform(n_samples=n_centers)
+
+        for i in range(max_iter):
+            dists = self.__intersets_distances(
+                centers, points, n_jobs=n_jobs, **joblib_kwargs
+            )
+
+            # assuming the use of 'flat' kernel
+            if kernel == "flat":
+                weights = np.ones_like(dists)
+
+            weights[dists > self.bandwidth] = 0.0
+            weights = weights / weights.sum(axis=1, keepdims=1)
+
+            points_to_average, nonzero_weights = [], []
+
+            for j in range(n_centers):
+                points_to_average += [
+                    points[np.where(weights[j, :] > 0)[0], :],
+                ]
+                nonzero_weights += [
+                    weights[j, :].nonzero()[0],
+                ]
+
+            # compute Frechet means in parallel
+            pool = joblib.Parallel(n_jobs=n_jobs, **joblib_kwargs)
+            out = pool(
+                pickable_mean(points_to_average[j], nonzero_weights[j])
+                for j in range(centers.shape[0])
+            )
+
+            new_centers = np.array(out)
+
+            displacements = [
+                self.metric.dist(centers[j], new_centers[j]) for j in range(n_centers)
+            ]
```

Comment:
The dist function works with lists of inputs. Are `centers[j]` and `new_centers[j]` single points? In this case you can use `self.metric.dist(centers, new_centers)` directly